### PR TITLE
Add online training with TD-MPC as proof of concept

### PR DESCRIPTION
**Pull request by alexander-soare**
## What this does

- Adds online training infrastructure in lerobot/scripts/online_training_helpers.py
- Adds online training loop in train.py
- Updates TD-MPC with minor fixes and refactors, and parallelizes the MPC.

TODO:
- [x] Add unit tests for `OnlineBuffer`.
- [x] Try training tdmpc_pusht with an offline training component.
  - Run is here https://wandb.ai/alexander-soare/lerobot/runs/fnhp9don. The results weren't better, but it got to a good overlap metric with way less training steps. 
- [x] Try training TD-MPC with PushT images.
  - Run is here https://wandb.ai/alexander-soare/lerobot/runs/ifmrwqjf. I used the same hparams as with `pusht_keypoints`. It didn't work out. I'm okay with merging this PR just knowing that it at least runs without an exception.  
- [x] Decide how to default the TD-MPC configuration.
- [x] Add last observation
  - Run is here https://wandb.ai/alexander-soare/lerobot/runs/jl5opn7z?nw=nwuseralexandersoare. The results aren't obviously better but when you smooth the training curves it looks like they might be a tiny bit better. I'm going to work this in.

## NOTE

This PR introduces a lot of duplication / technical debt:
- `OnlineBuffer` duplicates a lot of the interface of `LeRobotDataset`.
- `compute_sampler_weights` duplicates the functionality of `EpisodeAwareSampler`.

## How it was tested

Apart from CI:

- I tried training `tdmpc_pushtq_keypoints` with online only training and got solid results: https://wandb.ai/alexander-soare/lerobot/runs/1nyvnqyb?nw=nwuseralexandersoare
- Train `tdmpc_pusht` (image observations). The policy didn't learn anything useful but at least training ran without crashing: https://wandb.ai/alexander-soare/lerobot/runs/ifmrwqjf?nw=nwuseralexandersoare . I'm assuming hparams need to be tuned.
- Train `tdmpc_pusht_keypoints` with offline data in the mix. https://wandb.ai/alexander-soare/lerobot/runs/fnhp9don the maximum success metric was reached with less training iterations, but this metric was about the same as without offline training.
  
## How to checkout & try? (for the reviewer)

Try evaluating the model I trained with `python lerobot/scripts/eval.py -p alexandersoare/tdmpc_pusht_keypoints`. In the `eval_info.json` you should see the following aggregated results:

```json
{
    "avg_sum_reward": 177.15609529091438,
    "avg_max_reward": 0.9257891016804382,
    "pc_success": 34.0,
    "eval_s": 31.997821807861328,
    "eval_ep_s": 0.6399564504623413
}
```

### CODE REVIEW

**Comment by alexander-soare** [comment]
@Cadene @michel-aractingi @marinabar there are a couple of things in this PR that we should watch out for:

1. `OnlineBuffer` duplicates a lot of what happens in `LeRobotDataset`. I discussed with @Cadene that ultimately we want to have a `LeRobotDataset` class which is decoupled from the underlying storage mechanism, or at least only assumes something simple of the form `dict[str, torch.tensor]` or `dict[str, np.ndarray]`. This bit of technical debt is catching up with us and I think we'll have to deal with it next.

2. This PR re-introduced online training into the main training script, and adds a bunch of online training params in the hydra config. I've tried to separate the online training neatly from the offline training. I'm _okay_ with it now. Dunno about you both. Happy to consider alternatives.

3. I have added the ability to do online rollouts asynchronously so parallelize with training. In the end I didn't use it to train PushT because it was less efficient when using rollout batch size of 10. But for rollout batch size of 1 it _can_ be faster. I left the feature in for now but can remove it if we determine it's too complex. @marinabar can you please take a look at the train script and consider it in light of your PR with `accelerate`? I think that we should probably keep `accelerate` out of online training for now, but I still like the ability to use `amp`.

File: lerobot/scripts/online_training_helpers.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+            "episode_index": {"dtype": np.dtype("int64"), "shape": (buffer_capacity,)},
+            "timestamp": {"dtype": np.dtype("float64"), "shape": (buffer_capacity,)},
+            **{
+                k: {"dtype": v["dtype"], "shape": (buffer_capacity, *v["shape"])}
+                for k, v in data_spec.items()
+            },
+        }
+        return data_spec
+
+    def add_data(self, data: dict[str, np.ndarray]):
```
Comment (Cadene) [inline lerobot/scripts/online_training_helpers.py:131]: In our use case, is it frame, a trajectory, an episode?

From my understanding, it's an episode, but we can support all three. Could be nice to add a comment maybe?

File: lerobot/scripts/online_training_helpers.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+            "_occupancy_mask": {"dtype": np.dtype("?"), "shape": (buffer_capacity,)},
+            "index": {"dtype": np.dtype("int64"), "shape": (buffer_capacity,)},
+            "frame_index": {"dtype": np.dtype("int64"), "shape": (buffer_capacity,)},
+            "episode_index": {"dtype": np.dtype("int64"), "shape": (buffer_capacity,)},
+            "timestamp": {"dtype": np.dtype("float64"), "shape": (buffer_capacity,)},
+            **{
+                k: {"dtype": v["dtype"], "shape": (buffer_capacity, *v["shape"])}
+                for k, v in data_spec.items()
+            },
+        }
```
Comment (Cadene) [inline lerobot/scripts/online_training_helpers.py:128]: nit: could we use `new_data_spec = {}` and avoid one liner syntax?

File: lerobot/scripts/train.py [hunk 3]
```diff
@@ -420,13 +429,188 @@ def evaluate_and_checkpoint_if_needed(step):
...
+    online_env = make_env(cfg, n_envs=cfg.training.online_rollout_batch_size)
+    resolve_delta_timestamps(cfg)
+    online_buffer_path = logger.log_dir / "online_buffer"
+    if cfg.resume and not online_buffer_path.exists():
+        # If we are resuming a run, we default to the data shapes and buffer capacity from the saved online
+        # buffer.
+        logging.warning(
+            "When online training is resumed we use the latest online buffer, which is not the necessarily "
+            "the one that was present at the time of the provided checkpoint!"
+        )
```
Comment (Cadene) [inline lerobot/scripts/train.py:453]: I am not sure to understand. Could you clarify?

File: lerobot/scripts/train.py [hunk 4]
```diff
@@ -420,13 +429,188 @@ def evaluate_and_checkpoint_if_needed(step):
...
+    online_dataset = OnlineBuffer(
+        online_buffer_path,
+        data_spec={
+            **{k: {"shape": v, "dtype": np.dtype("float32")} for k, v in policy.config.input_shapes.items()},
+            **{k: {"shape": v, "dtype": np.dtype("float32")} for k, v in policy.config.output_shapes.items()},
```
Comment (Cadene) [inline lerobot/scripts/train.py:458]: This onliner make sense to me

File: lerobot/scripts/train.py [hunk 5]
```diff
@@ -420,13 +429,188 @@ def evaluate_and_checkpoint_if_needed(step):
...
+        pin_memory=device.type != "cpu",
+        drop_last=True,
+    )
+    dl_iter = cycle(dataloader)
+
+    # Lock and thread pool executor for asynchronous online rollouts. When asynchronous mode is disabled,
+    # these are still used but effectively do nothing.
+    lock = Lock()
+    executor = ThreadPoolExecutor(max_workers=1)
+
```
Comment (Cadene) [inline lerobot/scripts/train.py:507]: Impressive implementation. How slow sync is versus async? Could be worth adding an info here.

Why is max_workers set to 1?

File: lerobot/scripts/online_training_helpers.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        return self._item_to_tensors(item)
+
+
+def update_online_buffer(
+    online_dataset: OnlineBuffer,
+    concat_dataset: torch.utils.data.ConcatDataset,
+    sampler: torch.utils.data.WeightedRandomSampler,
+    new_data_dict: dict[str, torch.Tensor],
+    online_sampling_ratio: float,
+):
```
Comment (Cadene) [inline lerobot/scripts/online_training_helpers.py:262]: The best way to convey that we update inplace online_dataset, concat_dataset, sampler, might be to return them even tho it's ugly. What do you think?

**Review by Cadene — COMMENTED** [review]
Did a first pass (a bit shallow since I am tired). Really nice work. We could try to simplify the OnlineBuffer a bit to be a bit more pythonic. But overall looks mergeable soon after adding some tests!

Thanks Alexander 🙏

File: lerobot/scripts/train.py (hunk 5 above)
Comment (alexander-soare) [reply lerobot/scripts/train.py:507]: Regarding your question about `max_workers`, I hope this comment helps:

```
# Note: 1 worker because we only ever want to run one set of online rollouts at a time.
# Batch parallelization of rollouts is handled within the job.
```

File: lerobot/scripts/online_training_helpers.py (hunk 6 above)
Comment (alexander-soare) [reply lerobot/scripts/online_training_helpers.py]: FYI this is no longer relevant as I decided to switch this for a function that computes sampler weights. The rest of the logic went to `train.py`. So now nothing is being modified in place.

File: lerobot/common/policies/tdmpc/modeling_tdmpc.py [hunk 7]
```diff
@@ -19,14 +19,10 @@
 The comments in this code may sometimes refer to these references:
     TD-MPC paper: Temporal Difference Learning for Model Predictive Control (https://arxiv.org/abs/2203.04955)
     FOWM paper: Finetuning Offline World Models in the Real World (https://arxiv.org/abs/2310.16029)
-
-TODO(alexander-soare): Make rollout work for batch sizes larger than 1.
-TODO(alexander-soare): Use batch-first throughout.
```
Comment (Cadene) [inline lerobot/common/policies/tdmpc/modeling_tdmpc.py:24]: Awesome!

File: lerobot/configs/policy/tdmpc.yaml [hunk 8]
```diff
@@ -4,19 +4,27 @@ seed: 1
...
+  log_freq: 100
+
+  online_steps: 50000
+  online_rollout_n_episodes: 1
+  online_rollout_batch_size: 1
+  # Note: in FOWM this is actually dynamically set to match exactly the length of the last sampled episode.
+  online_steps_between_rollouts: 50
+  online_sampling_ratio: 0.5
+  online_env_seed: 10000
+  # FOWM Push uses 10000 here which is approx 400 of their episodes worth. We'll also go for 400 episodes.
```
Comment (Cadene) [inline lerobot/configs/policy/tdmpc.yaml:25]: I didnt understand this comment :')

File: lerobot/configs/policy/tdmpc.yaml [hunk 9]
```diff
@@ -4,19 +4,27 @@ seed: 1
...
   grad_clip_norm: 10.0
   lr: 3e-4
 
+  eval_freq: 5000
+  log_freq: 100
+
+  online_steps: 50000
+  online_rollout_n_episodes: 1
+  online_rollout_batch_size: 1
+  # Note: in FOWM this is actually dynamically set to match exactly the length of the last sampled episode.
```
Comment (Cadene) [inline lerobot/configs/policy/tdmpc.yaml:21]: What is dynamical in FOWM?

File: lerobot/scripts/online_training_helpers.py [hunk 10]
```diff
@@ -0,0 +1,292 @@
+"""A collection of helper functions and classes for the online training loop in train.py"""
```
[the remark this answers is not in the document]
Comment (Cadene) [reply lerobot/scripts/online_training_helpers.py]: I would prefer to only have scripts in the `scripts` folder. Could we please add it it `common`?
We can always move the content in other parts.

Also can we rename `online_training.py` ?

File: lerobot/scripts/train.py (hunk 3 above)
Comment (Cadene) [reply lerobot/scripts/train.py]: Sounds good!

typo: mechansims

**Review by Cadene — APPROVED** [review]
Left a few comments that might need to be addressed. Approved to unblock.
Thanks for this important PR + tests :)

File: lerobot/configs/policy/tdmpc.yaml (hunk 8 above)
Comment (alexander-soare) [reply lerobot/configs/policy/tdmpc.yaml:25]: Updated.

File: lerobot/common/policies/tdmpc/modeling_tdmpc.py (hunk 7 above)
Comment (alexander-soare) [reply lerobot/common/policies/tdmpc/modeling_tdmpc.py:24]: Almost. I did the first TODO but I determined that the second TODO should not be done.

File: lerobot/scripts/online_training_helpers.py [hunk 11]
```diff
@@ -0,0 +1,369 @@
+"""A collection of helper functions and classes for the online training loop in train.py
+
+Note to maintainers: This duplicates some logic from LeRobotDataset and EpisodeAwareSampler. We might consider
+converging to one approach when/if this functionality moves to core.
```
Comment (michel-aractingi) [inline lerobot/scripts/online_training_helpers.py:4]: Perhaps we can expand abit around we used numpy here?

File: lerobot/scripts/online_training_helpers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (alexander-soare) [reply lerobot/scripts/online_training_helpers.py]: Done
